### PR TITLE
Support int to decimal conversion

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Related to #289

### CODE REVIEW

File: runtime/decimal.c [hunk 1]
```diff
@@ -167,6 +167,20 @@ double _bal_decimal_to_float(TaggedPtr tp) {
     return strtod(dblStr, NULL);
 }
 
+TaggedPtr _bal_decimal_from_int(int64_t val) {
+    decQuad d;
+    if (val <= INT32_MAX && val >= INT32_MIN) {
+        decQuadFromInt32(&d, val);
+        return createDecimal(&d);
+    }
+    char intStr[21]; // max length + 1
```
Comment (jclark) [inline runtime/decimal.c]: How do you know 21 is enough? Use an expression that shows it.

File: runtime/decimal.c [hunk 2]
```diff
@@ -167,6 +167,20 @@ double _bal_decimal_to_float(TaggedPtr tp) {
     return strtod(dblStr, NULL);
 }
 
+TaggedPtr _bal_decimal_from_int(int64_t val) {
+    decQuad d;
+    if (val <= INT32_MAX && val >= INT32_MIN) {
```
Comment (jclark) [inline runtime/decimal.c]: Write this as in maths like:

```
INT32_MIN <= val && val <= INT32_MAX
```

File: runtime/decimal.c [hunk 3]
```diff
@@ -167,6 +167,20 @@ double _bal_decimal_to_float(TaggedPtr tp) {
...
+        decQuadFromInt32(&d, val);
```
Comment (jclark) [inline runtime/decimal.c]: Explicit cast `val` to `int32_t`.

**Review by jclark — CHANGES_REQUESTED** [review]

File: runtime/decimal.c (hunk 1 above)
Comment (KavinduZoysa) [reply runtime/decimal.c]: Added a comment

File: runtime/decimal.c (hunk 2 above)
Comment (KavinduZoysa) [reply runtime/decimal.c]: Changed

File: runtime/decimal.c (hunk 3 above)
Comment (KavinduZoysa) [reply runtime/decimal.c]: Changed

File: runtime/decimal.c [hunk 4]
```diff
@@ -167,6 +167,22 @@ double _bal_decimal_to_float(TaggedPtr tp) {
...
 
+TaggedPtr _bal_decimal_from_int(int64_t val) {
+    decQuad d;
+    if (INT32_MIN <= val && val <= INT32_MAX) {
+        decQuadFromInt32(&d, (int32_t)val);
+        return createDecimal(&d);
+    }
+    // The INT64_MIN(-9223372036854775808) has the maximum string length.
+    // It has 20 characters. Adding +1 for NULL terminator.
+    char intStr[21];
```
Comment (jclark) [inline runtime/decimal.c]: I want the code to be clearly correct. How do I see how many digits are in `9223372036854775808`? How do I know that is -2^63?

What I want is `21` replaced by an expression that is clearly >=the number of digits needed.

For example, I know that one decimal digit can represents at least 3 binary bits.

So I know that ceiling(64/3) + 1 for sign + 1 for zero is enough.

Or you can have a macro that does sizeof on the stringification of INT64_MIN.

**Review by jclark — CHANGES_REQUESTED** [review]

File: runtime/decimal.c (hunk 4 above)
Comment (KavinduZoysa) [reply runtime/decimal.c]: Changed

File: runtime/decimal.c [hunk 5]
```diff
@@ -167,6 +171,20 @@ double _bal_decimal_to_float(TaggedPtr tp) {
     return strtod(dblStr, NULL);
 }
 
+TaggedPtr _bal_decimal_from_int(int64_t val) {
+    decQuad d;
+    if (INT32_MIN <= val && val <= INT32_MAX) {
+        decQuadFromInt32(&d, (int32_t)val);
+        return createDecimal(&d);
+    }
+    char intStr[INT64_MAX_LEN];
```
Comment (jclark) [inline runtime/decimal.c]: Better to do `else` here and then do `createDecimal` once.

File: runtime/decimal.c [hunk 6]
```diff
@@ -9,6 +9,10 @@ typedef GC decQuad *DecimalPtr;
 // XXX need to investigate DEC_Underflow
 #define DECIMAL_STATUS_FAIL DEC_Errors
 
+#define STR_CONVERT(x) #x
+#define STR(x) STR_CONVERT(x)
+#define INT64_MAX_LEN sizeof(STR(INT64_MIN))
```
Comment (jclark) [inline runtime/decimal.c]: Move these defines to just before you use INT64_MAX_LEN.

**Review by jclark — CHANGES_REQUESTED** [review]

File: runtime/decimal.c (hunk 5 above)
Comment (KavinduZoysa) [reply runtime/decimal.c]: Changed

File: runtime/decimal.c (hunk 6 above)
Comment (KavinduZoysa) [reply runtime/decimal.c]: Changed

**Pull request merged**